### DISCUSSION TITLE
Non update-able columns?

### DISCUSSION THREAD

**Discussion by Riuen** [adonisjs/core#3362] [Help]
Does a feature currently exist which allows you to essentially mark a column as not update-able? In other words, those fields are excluded from update operations whenever the save method on the model is called.

Eg:
`class Foo extends BaseModel {
@column({ isPrimary: true })
public id: number

@column()
public field1: string;

@column({updateable: false})
public field2: string;
}
`
In the example above, field2 will always be ignored when updating the table.

**Comment by Gredys** [comment]
I was searching for something like this a few days already. 
I just need custom fields in model that can exist only on server side and not on db.

Usage example.
We have user model with fields like [id,user,...]. They are saved to db and everything is ok.
For example we have some runtime field, like token, for user that can only exists while server is running.
Saving such runtime data in db is redundant and bad in general.
In current situation (that I found) we forced to create another class instance to hold relations to the user id and custom data.
But what would be awesome - just allow creating custom fields that are only server side and not saved to db, and can be populated (for example) on afterFetch and after loading instance from db.
In this way custom logic related to user model would be simplified to mininun without creating numeroues classes and instances just to hold additional data.

P.S. If we already have such feature - sorry. Please point me with url to doc. Thanks.

**Reply by thetutlage** [reply]
Just define a property on the class and don't put `@column` decorator on it

**Reply by Gredys** [reply]
Yeah its possible to do so. But it doesn't work with serialization (at least for me), which is required to utilize field restrictions and so on.
To work with serialization I am creating computed property and just returning my custom field and settings serializeAs with proper name.
But its will create mess in code if u have many fields, which is sign of bad code style.
Thats why it would be better to allow using `@column` and just tell it to ignore db but allow usage of serialization and etc.

**Reply by Gredys** [reply]
Current example.

Can't use before/after fetch. If field is null by default changing it in callback doesn't change value and field still null after serialiaztion.
Forced to use `@computed` to display value after serialization.

It would be better just declare field as `@column` and tell it that field is `custom`.
Idk what about before/after fetch. Mby just fix usage so value would be changed and let users to decide what to do.

```
  public token: string | null = uuidv4()

  // Not working for some reason.
  //     Nor in beforeFetch nor in afterFetch
  // @afterFetch()
  // static setToken(user: User) {
  //   // TEMP: Generate UUID
  //   user.token = uuidv4()
  // }

  // Allows to see field in serialization
  @computed({serializeAs: "token"})
  get getToken() {
    return this.token
  }
  ```

**Reply by Gredys** [reply]
Comparing to what it could be (naming is just for example). A lot less code and mess.

Variant #1 (just declare):
```
  @column({custom: true})
  declare token: string

  // Fill multiple custom fields with values
  @afterFetch()
  static initCustom(user: User) {
    user.token = uuidv4()
    // ...
  }
```
Variant #2 (declare and set):
```
  @column({custom: () => {
    return uuidv4()
  }})
  declare token: string
  ```

**Reply by thetutlage** [reply]
How about doing this?

```ts
  @computed()
  declare token: string

  // Fill multiple custom fields with values
  @afterFetch()
  static initCustom(user: User) {
    user.token = uuidv4()
    // ...
  }
```

**Reply by Gredys** [reply]
It looks good also, but for some reason `afterFetch` doesn't call at all.
I am using `User.findBy` and i've set 3 console prints.
2 inside afterFetch (before and after value set) and 1 after findBy.
The last one got called only.

**Reply by thetutlage** [reply]
After fetch is called after you fetch more than one row from the database. The `findBy` method returns a single row and you must use `afterFind` hook in that case.
